src/opera/commands/diff.py

```diff
@@ -0,0 +1,219 @@
+
+        assert False, "BUG - invalid format"
+
+
+def get_template(storage):
+    if storage.exists("inputs"):
+        inputs = storage.read_json("inputs")
+    else:
+        inputs = {}
+
+    if storage.exists("root_file"):
+        service_template = storage.read("root_file")
+        workdir = str(Path.cwd())
+
+        if storage.exists("csars"):
+            csar_dir = Path(storage.path) / "csars" / "csar"
+            workdir = str(csar_dir)
+            ast = tosca.load(Path(csar_dir),
+                             PurePath(service_template).relative_to(csar_dir))
+        else:
+            ast = tosca.load(Path.cwd(), PurePath(service_template))
+
+        template = ast.get_template(inputs)
+        return template, workdir
+    else:
+        return None, None
+
+
+def diff_templates(service_template_old: str, workdir_old,
+                   inputs_old: typing.Optional[dict],
+                   service_template_new: str, workdir_new,
+                   inputs_new: typing.Optional[dict],
+                   template_comparer: TemplateComparer, verbose_mode: bool):
+    """
+    :raises ParseError:
+    :raises DataError:
+    """
+    if inputs_new is None:
+        inputs_new = {}
+
+    if inputs_old is None:
+        inputs_old = {}
+
+    ast_old = tosca.load(workdir_old, PurePath(service_template_old))
+    ast_new = tosca.load(workdir_new, PurePath(service_template_new))
+
+    template_old = ast_old.get_template(inputs_old)
+    template_new = ast_new.get_template(inputs_new)
+    context = TemplateContext(template_old,
+                              template_new,
+                              workdir_old,
+                              workdir_new)
+
+    equal, diff = template_comparer.compare_service_template(template_old,
+                                                             template_new,
+                                                             context)
+    return diff
+
+
+def diff_instances(storage_old: Storage, storage_new: Storage,
+                   template_comparer: TemplateComparer,
+                   topology_comparer: InstanceComparer,
+                   verbose_mode: bool):
+    """
+    :raises ParseError:
+    :raises DataError:
+    """
+
+    template_old, workdir_old = get_template(storage_old)
+    template_new, workdir_new = get_template(storage_new)
+    topology_old = template_old.instantiate(storage_old)
+    topology_new = template_new.instantiate(storage_new)
+
+    context = TemplateContext(template_old,
+                              template_new,
+                              workdir_old,
+                              workdir_new)
+
+    equal, diff = template_comparer.compare_service_template(template_old,
+                                                             template_new,
+                                                             context)
+
+    equal, diff = topology_comparer.compare_topology_template(topology_old,
```

Comment:
Does this compare instances? Would then be better `topology_comparer.compare_instances`?

Also I tried to deploy `service.yaml` from `examples/hello/` and then I changed service template by adding one new input (that is not used anywhere):

```yaml
topology_template:
  inputs:
    marker:
      type: string
      default: default-marker
    input:
      type: string
      default: input
```

I also created `inputs.yaml`:

```yaml
marker: test
input: test-input
```

Then I executed `opera diff` and got the following result:

```console
(.venv) anzoman@ubuntu:~/Desktop/xopera-opera$ opera diff -i examples/hello/inputs.yaml examples/hello/service.yaml --format yaml
nodes:
  hello:
    interfaces:
      Standard:
        operations:
          create:
            inputs:
              marker:
              - default-marker
              - test
          delete:
            inputs:
              marker:
              - default-marker
              - test
```

So, I the difference says that I changed the value for `marker` input, but it doesn't say that I added a new input. Is it meant to be that only used values are seen on diff? Or should there be one additional yaml keyname? Like this:

```yaml
inputs:
  test:
  - test-input
```

Reply:
The idea that we do not compare all inputs, but only inputs that have been reflected in some node properties. Basically if you just define a new input that does not affect any node there is no need for deployment update.

Reply:
Ok, this sounds reasonable. We could then add comparing all inputs in the future if we will really need that functionallity.